### ISSUE TITLE
trigger: no-flow triggered tasks should always merge whether still in the pool or not

### ISSUE THREAD

**Issue by oliver-sanders** [cylc/cylc-flow#4657]
This is a condensation of concerns raised with the `cylc trigger` logic on #4651 a different formulation of which can be found in https://github.com/cylc/cylc-flow/issues/4653.

**Outline:**

* When a task that has not yet run is triggered via `cylc trigger` [without `--reflow`] it is not assigned to a flow but will merge with the first flow that catches it.
* If this task is still in the pool by the time the scheduler window catches up with it it merges into the flow that caught it.
* If the task has exited the pool before the scheduler window catches up with it the task is forgotten and does not merge.

*Is this analysis correct?* **[Yes](https://github.com/cylc/cylc-flow/issues/4657#issuecomment-1032063777)**

**Problem:**

* This creates a nasty situation where a task is either run once or twice, depending on the evolution of the scheduler during the runtime of the task.
* This makes it hard for users to understand the consequences of the trigger before they issue it.
* Tasks triggered ahead of the scheduler window will always be run twice, which is highly unlikely to be what the user wanted.

*Do we agree that this is a problem?* **[No](https://github.com/cylc/cylc-flow/issues/4657#issuecomment-1032063777)**

**Desired Behaviour:**

* Either the triggered task should always merge, or it should never merge, it shouldn't be able to go either way.
* Task triggering is not supposed to be a reflow feature, users should be able to use it without understanding reflow.

*Do we agree that this is the desired behaviour?* **[No](https://github.com/cylc/cylc-flow/issues/4657#issuecomment-1032438108)**

**Comment by hjoliver** [comment]
> Is this analysis correct? Is it missing anything?

Correct!

> This creates a nasty situation where a task is either run once or twice, depending on the evolution of the scheduler during the runtime of the task.

That's not nasty. It's a natural consequence of flow merging (which is a necessity until flow number is in the UID). Consider two ongoing flows, one behind the other.  Every task ahead of the second flow (the one that is catching up) will either run twice or once, depending on exactly when/if catch-up occurs. 

 > This makes it hard for users to understand the consequences of the trigger before they issue it.

If you trigger a task ahead of the main flow:
- it will only run once if the main flow catches up before it finishes
   - (because Cylc can't support two instances of the same `point/name` active at once)
- otherwise it will run twice (once when you trigger it, once when the main flow passes over it)
  - (unless we implement `cylc trigger --flow=1`, then the scheduler will know it already ran in this flow #4653) 
  
That's not hard to understand is it?

> Tasks triggered ahead of the scheduler window will always be run twice,

Yes (again, unless we implement `cylc trigger --flow=1`)

> which is highly unlikely to be what the user wanted.

My take was, if the dependencies encoded in the graph are meaningful then it seems unlikely that triggering future tasks (as part of the flow) is going to be wanted much. Which is not to say never, so we should do #4653 

But I do know it's quite common to test changes to tasks by manually triggering individual tasks in a running workflow, as any easy cheat (c.f. deploying a new instance of the whole workflow just to do that). And if doing that, it's best to trigger a future or past instance to avoid any confusion or clash with the running flow.  (This is essentially the old `cylc submit` use case).

**Comment by hjoliver** [comment]
Regarding the title:
>  no-flow triggered tasks should always merge whether still in the pool or not

"Flow merge" has so far meant what happens (necessarily, until the UID has flow in it) when two instances of the same task, from different flows, want to be in the active pool at the same time.

By that definition, flow merging can't occur "whether still in the pool or not".  

So presumably you mean (as I've assumed above) that the triggered task should be be triggered as part of the main flow, rather than as a "no flow" task.

**Comment by hjoliver** [comment]
Long story short, `cylc trigger --flow=` will solve the problem?

It remains to be seen if we can automatically assign the flow number in some circumstances (e.g. when there's only one flow present).

**Comment by hjoliver** [comment]
> Desired Behaviour:

>    Either the triggered task should always merge, or it should never merge, it shouldn't be able to go either way.

"never merge" is not possible, because of the UID problem.  If a task in a flow catches up to another instance of itself in n=0 (whether that instance is flow or no-flow) we **have to** merge them.

>   Task triggering is not supposed to be a reflow feature, users should be able to use it without understanding reflow.

They way I look at it, it's not: 
- Triggering without `--reflow` is independent of flows **unless** the main flow catches up while the triggered task is still running, and we can't avoid that (and it's not hard to understand - do users really want two instances of the exact same task running at the same time?)
- Triggering ahead with `--reflow` (whether as part of the origin flow, via `--flow=` or not) is by definition using the reflow concept (in the sense of manually triggering a task from which the flow continues on).

- understand that the main flow can merge into the triggered task if it catches up while its still running, which

**Comment by oliver-sanders** [comment]
> That's not hard to understand is it?

I think that is quite hard to understand, at least I find it confusing and surprising, hence I reported it as a bug. If the behaviour is unintuitive to someone who works on the project...

This just feels needlessly complicated to me. Users should not need to understand reflows or merging conditions to use `cylc trigger` which should be a standalone concept.

> My take was, if the dependencies encoded in the graph are meaningful then it seems unlikely that triggering future tasks as part of the main flow is going to be wanted much

Perfectly reasonable, however, dependencies are not always meaningful and not all tasks have dependencies (the original reported issue and the orphan task use cases I contributed).

If a user is triggering a task with dependencies ahead of time and that task is going to stand a reasonable chance of doing something meaningful then it's reasonable to presume the dependencies aren't critical to its functioning.

> "Always merge" is not possible either

Pretty sure it is possible and probably quite easy, happy to implement.

No-flow tasks sit outside of the reflow model and as such are special anomalies.

It doesn't matter whether a no-flow task is still in the pool or not we can easily implement the logic to merge it with the first flow that catches it. We are already performing a DB check (almost) every time we spawn a task, we can easily check if it's a no-flow task and apply the merge condition at a later date (which I did on my original branch).

> Experimenting with task modifications the lazy way: by triggering "out of the way" tasks in your running workflow instead of deploying another instance of the whole workflow

Reasonable use case, however, I think that's something you should have to ask for, not the default behaviour.

**Comment by hjoliver** [comment]
BTW, Cylc 7 users would presumably expect "no-flow" manual triggering. That's what they get, unless they happen to trigger a task whose downstream children have already been pre-spawned into the task pool.

**Comment by hjoliver** [comment]
> If a user is triggering a task with dependencies ahead of time and that task is going to stand a reasonable chance of doing something meaningful then it's reasonable to presume the dependencies aren't critical to its functioning.

True, that's why I said "not wanted much" (as opposed to "never" :grin: )


> See also https://github.com/cylc/cylc-flow/issues/3748. [Retire task pool as a user-facing concept]
> 
> I think partially satisfied prerequisites and incomplete outputs might further convolute this?

Yes. Essentially it's complicated by waiting tasks that have to be assigned a flow already. However, this is still beyond simple compared to the nasty Cylc 7 task pool concept. Here, the `n=0` window contains active tasks, incomplete tasks, and "active waiting" tasks, the latter being simply "tasks that are ready to run but are held back for now by a limiting mechanism" (oh, and partially satisfied tasks - which also have to have a flow assigned already, as the prerequisites are flow-dependent).

**Comment by hjoliver** [comment]
OK I'm wondering if this is where our respective wires got crossed:

I've always thought of reflow as the ability to start multiple independent flows in different parts of the graph, whether past or future relative to the ongoing (or stopped) initial flow.
 - with the proviso that (until #4419 flows in UID) we can't support total independence and therefore (for now) have to merge if the same task ID gets spawned at the same time in `n=0`
 - and a triggered no-flow task is just a single-task subset of a flow, which doesn't need a flow number because it doesn't spawn children
- but if you want to trigger a future task as part of an existing flow, you just have to use `trigger --flow=` (not implemented yet, but the `set-outputs` case is).

Whereas you seem to be suggesting different flows should always merge if they cover the same part of the graph - not just if they clash in `n=0` ???  

Or are you only suggesting always-merge behaviour for triggered no-flow tasks?  Otherwise this would seemingly prevent triggering of multiple flows at all, if they cover the same parts of the graph at different times, which would rule out what's likely the primary use case for reflow, i.e. regeneration of some past products. 

Or maybe I've misunderstood ... I'm trying to digest this while sitting in a conference!!

**Comment by oliver-sanders** [comment]
I think we are on the same page!

> I've always thought of reflow as the ability to start multiple independent flows in different parts of the graph, whether past or future relative to the ongoing (or stopped) initial flow.

I agree.

The confusion may have arisen because I have likened no-flow tasks to a special case of reflow where the spawning logic is disabled.

> Or do you mean only for a flow that overruns a triggered no-flow task (that finished before the flow got there)?? (which is after all is the topic of this issue!).

Yes.

> If the latter, I don't have a problem with that, it's just a minor tweak to the model I've just described (and in fact it's the same as the last bullet point except that you presumably don't want to have to use --flow= by default?

Yes.

**Comment by oliver-sanders** [comment]
Here are some examples of what I would consider to be the "desired behaviour" to make sure we're on the same page:

**In this example:**

```
a(running) => b(runahead) => c => d => e
```

* If I triggered "b", I would expect "b" to run once and only once.
* If I triggered "c", I would expect "c" to run once and only once.
* If I triggered "d", I would expect "d" to run once and only once.
* If I triggered "e", I would expect "e" to run once and only once.

**In this example:**

```
1/a(running) => 2/a(runahead) => 3/a => 4/a => 5/a
```

* If I triggered "2/a", I would expect "2/a" to run once and only once.
* If I triggered "3/a", I would expect "3/a" to run once and only once.
* If I triggered "4/a", I would expect "4/a" to run once and only once.
* If I triggered "5/a", I would expect "5/a" to run once and only once.

**In this example:**

```
1/a(running)
2/a(runahead)
3/a
4/a
5/a
```

* If I triggered "2/a", I would expect "2/a" to run once and only once.
* If I triggered "3/a", I would expect "3/a" to run once and only once.
* If I triggered "4/a", I would expect "4/a" to run once and only once.
* If I triggered "5/a", I would expect "5/a" to run once and only once.

**Comment by hjoliver** [comment]
OK I think I understand you now :tada:

I was worried about wider implications because I thought you meant the same for flows as for single triggered tasks (because to me, so far, a triggered task is the same as a flow, except that it doesn't flow onward from the original trigger point).

So I agree that we need to support flows merging with manually triggered future tasks (in your sense of "merge", not just an `n=0` merge as it has meant so far in SoD implementation an documentation).

The difference is, my intention was to allow this via `trigger --flow=`, whereas you think it is the  desirable default behaviour.

I certainly agree that could be done as per your examples just above, if all we have is the original flow and manually triggered tasks out in front of.  But will it work when there are multiple flows in play? 

My way `cylc trigger --flow=` will always work in the following sense: if you have to specify the flow number, that particular flow will not repeat-run the future task when it reaches that place in the graph. But other flows will.  

What if, say, the original flow has already passed by our location in the graph, and now a second flow is approaching, and we trigger a task that is in the future with respect to the second flow, but in the past with respect to the original flow? How does the second flow know that the triggered task was only meant to belong to the first flow, even though we did use explicitly state that with `trigger --flow=`?

Presumably you are not suggesting that we record in the DB that a task was triggered as a no-flow, and if so *any* subsequent flow will not repeat-run it?

**Comment by hjoliver** [comment]
Thinking of how this might work:

If I manually trigger a task (without the `--reflow` option):
- it gets tagged in the DB with an explicit zero flow number as a no-flow task 
- the first flow that overruns the task takes possession of it, by replacing the zero with its flow number (retroactive DB update)
- then to subsequent flows, the task will look like it always belonged to that first flow

Note we have to use an explicit no-flow indicator (zero flow number) to the two-flow scenario above. Because the triggered task might not be a "future task" with respect to all flows (i.e. it might have already run before, in another flow).
 
What do you think?

**Comment by hjoliver** [comment]
> (although I wasn't considering preserving an explicit no-flow number so {1} rather than {0, 1})

OK, yeah we can stick with the current no-flow-number approach, because the upcoming flow queries the most recent submit number before spawning.

**Comment by hjoliver** [comment]
I don't quite understand your example 3. Assuming `c#1` means job-submit no.`1` of task `c`, we can't have `c#1(running)` and `c#2(running)` at the same time, as in two of your bullet-point lists, because they can't coexist in the scheduler task pool.

**Comment by hjoliver** [comment]
For anyone trying to follow this:

- We're on the same page with regard to the functionality: we need to support triggering a future task as part of an upcoming flow, so that the triggered task doesn't get re-run when the flow gets to that place in the graph 
- We just need to nut out whether that can be done generally without having to use `trigger --flow=`

**Comment by hjoliver** [comment]
The implicit approach has one interesting complication, which shows it would be good to have `trigger --flow` as option as well, even for non-"exotic" cases. 

Spawning "on demand" is local in the graph. i.e. each task spawns its own children, as part of its own flow, as it completes its  outputs. So:

1. `cylc trigger 3/c`
- `3/c` can't spawn its children as it runs because it doesn't belong to a flow yet
- we have to retroactively spawn on its outputs once the flow catches up and "possesses" it

2. `cylc trigger --flow=2 3/c` 
- `3/c` can spawn its children normally, as it runs
- the children may also be able to run (and flow on), depending, even before the main flow catches up 

Both approaches could be useful. I suppose it is reasonable to have 1. as the default though.

**Comment by hjoliver** [comment]
Another minor complication:

I think it's fair to say that when triggering a task in front of a flow, users will most often want the following flow not to rerun the task.

But the opposite is probably true for trigger a task behind a flow.  In this case, users are most likely wanting to rerun a task in the flow that passed by already, not to pre-run a task in a future flow that may not have even be started yet.

**Comment by hjoliver** [comment]
Seems to me we need this:
- `cylc trigger --rerun ID`
  - rerun a single task from previous flow; ignored by the following flow
- `cylc trigger --prerun ID`
  - prerun a task; the following flow will merge with it; retroactively spawn on its outputs
- `cylc trigger --prerun --flow=2 ID`
  - prerun a task as part of flow 2; children can be spawned as it runs

And, it is an error not give either `--prerun` or `--rerun`.

**Comment by oliver-sanders** [comment]
> So it seems to me we need something like this:

`--pre-run`, `--re-run` is an interesting idea, would be tempted to make one of the two the default.

I had expected the CLI to end up like this:

> `cylc trigger --re-run ID`

```
cylc trigger --flow=1 ID
```

> `cylc trigger --pre-run ID`

```
cylc trigger ID
```

> `cylc trigger --pre-run --flow=2 ID`

```
cylc trigger --flow=2 ID
```

**Comment by oliver-sanders** [comment]
It is done: https://github.com/cylc/cylc-flow/issues/4686
